### PR TITLE
feat: getTradeRate swapper endpoint

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR brings `getTradeRate()` as en endpoint
  - For ZRX, brings in the notion of a temporary "pseudo trade quote" so things are non-breaking for current `getTradeQuote` endpoints. The current implementation won't play nice with the split of trades and quotes, since our current endpoint is a mix of both (calls the inner `/price` fetcher, but assumes `quote` input, and does some gas calcs that assume wallet is present).
  - For Portals, CoW, Arb Bridge, Li.Fi and THOR we're talking a proper split between the two
- Gets rates instead of quotes if the `PublicTradeRoute` flag is on (which it will never be for the time being, including for dev, unless you explicitly set it to `true`)

## Issue (if applicable)

- contributes to (but doesn't consume) https://github.com/shapeshift/web/issues/7941 
- contributes to https://github.com/shapeshift/web/issues/7979

## Risk

> High Risk PRs Require 2 approvals

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

### Engineering

⚠️ Note: This inherently breaks execution with the `PublicTradeRoute` flag on, which is expected. The flag has been disabled not to break dev until this is fully wired up.

**Test all the below with the `PublicTradeRoute` flag off**

- Able to get quotes for (and execute) permit2 ZRX with permit2 flag 
- Able to get quotes for (and execute)  classic ZRX without permit2 flag 
- Able to get quotes for (and execute) Portals (note, this is a non-breaking change, hence the super hacky "quote then maybe estimate then maybe another dummy quote" flow is still present, to be removed when we wire this up with rates instead of quotes, with "TBA" for network fees)
- Able to get quotes (and execute) for CoW
- Able to get quotes (and execute) for Li.Fi
- Able to get quotes (and execute) for THOR

**Test all the below with the `PublicTradeRoute` flag on.**
How to confirm whether you're getting rates or quotes? Well, the easiest is to ensure that the newly added `getTradeRate` swapper endpoint is hit, with a debugger breakpoint. Alternatively/additionally, you can also ensure that sell/receive addies are not passed upstream as params when getting rates, e.g:

image
image

- for ZRX classic without permit2 flag, able to get  *rates* 
- for ZRX permit2 with permit2 flag, able to get *rates*
- for Portals, able to get *rates*
- for Arb Bridge, able to get *rates*
- for Li.Fi, able to get rates 
- for THOR, able to get rates 
- for CoW, able to get rates

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

- Able to get quotes for all swappers *and* execute them e2e. Do not try and test this with the `PublicTradeRoute` flag on yet, this is heavily WIP and for eng. only.

### `PublicTradeRoute` flag on

- for ZRX permit2 with permit2 flag, able to get *rates*

https://jam.dev/c/a7e4e489-5c61-4d70-a59f-ed7a9499d398

- for ZRX classic without permit2 flag, able to get  *rates* 

https://jam.dev/c/7a2cbb00-4b95-41e7-80cf-d90d274f0d00

- for Portals, able to get *rates* (note, network fee is TBD here)

https://jam.dev/c/2692f3d7-5634-4612-8e62-f49f1457aea6

- for Arb Bridge, able to get *rates*

https://jam.dev/c/e8ec9489-22d0-474a-9ad0-1925038e1bcc

- for Li.Fi, able to get rates 

https://jam.dev/c/23a87d54-561e-4241-bbbe-21781cacc088

- for THOR, able to get rates

https://jam.dev/c/b5075a07-7556-4051-892c-158aab62bbde

### Flag off

- Able to get (and execute) quotes for permit2 ZRX with permit2 flag 

https://jam.dev/c/b70a050e-b8a0-49f2-8cf2-479596f98e9d

- Able to get (and execute) quotes for classic ZRX without permit2 flag 

https://jam.dev/c/3fbf7210-fa66-4866-98a8-b4a581fcebcb

- Able to get (and execute) quotes for Portals

https://jam.dev/c/fcb4b5ae-fdd4-4c6b-be58-712af1f839dd

- Able to get quotes (and execute) for CoW

https://jam.dev/c/3293a6a6-f666-41ab-80d5-325ff34f5f21

- Able to get quotes (and execute) for Li.Fi

Unable to test because of current Li.Fi rate limits 

- for THOR, able to get (and execute) quotes  

https://jam.dev/c/608f187f-fa7e-492f-a47b-81df3d59d4c6

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Notes after discussing with woody:

- Let's try not to add a new RTK api slice and abuse the current `getTradeQuote` instead (assuming flag on) to get rates
- Execution may be borked with flag on if we do this, which is fine as we can just turn it off in base *and* dev
- Alternatively, can always duplicate everything tradeQuote with a new API/slice but that's going to be a huge lift / risks

tl;dr ditch all the boilerplate and console.logs validating this works and make it actually work i.e
- rates work with the flag on and execution probably borked (but ideally not) which is expected
- quotes are still happy with the flag off
- assess next steps re: state machine with @woodenfurniture once we're ready to merge this

File: src/state/apis/swapper/helpers/validateTradeQuote.ts
Comment (gomesalexandre) [inline src/state/apis/swapper/helpers/validateTradeQuote.ts:1]: review with whitespaces diff off

File: src/state/apis/swapper/swapperApi.ts [hunk 1]
```diff
@@ -75,14 +77,34 @@ export const swapperApi = createApi({
...
-        const quoteResult = await getTradeQuotes(
-          {
-            ...tradeQuoteInput,
-            affiliateBps,
-          },
-          swapperName,
-          swapperDeps,
-        )
+        const getQuoteResult = () => {
+          // Always get a trade rate with the flag enabled. This will break execution for the time being, but as discussed with @woodenfurniture
```
Comment (gomesalexandre) [inline src/state/apis/swapper/swapperApi.ts:81]: @woodenfurniture for vis since this inline tags no worky

File: .env.dev [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # feature flags
-REACT_APP_FEATURE_PUBLIC_TRADE_ROUTE=true
+REACT_APP_FEATURE_SOLANA=true
+REACT_APP_FEATURE_PHANTOM_WALLET=true
+REACT_APP_FEATURE_PUBLIC_TRADE_ROUTE=false
```
Comment (gomesalexandre) [inline .env.dev:4]: one too deep on the mergefix

**Review by woodenfurniture — CHANGES_REQUESTED** [review]
# Testing notes

❓ Metamask generally very flaky with signing, rarely pops up the signature modal in the extension, but this appears to affect prod too. Basically unable to test with metamask on any chain as metamask is not allowing the signing of txs.

✅ LiFi


✅ 0x with Permit2


✅ Portals


✅ Cow

❌ When connected to metamask (without snaps), user is able to select RUNE on THORChain but unable to get a rate. In prod this is not possible.


❌ Sitting on the screen above waiting for a re-fetch, the loading state gets stuck on.


Will test rates on second pass! 🙏

**Review by woodenfurniture — APPROVED** [review]
Confirmed issues raised are with flag off and are expected 🎉 LFGTM

File: src/pages/RFOX/components/Stake/Bridge/hooks/useRfoxBridge.ts [hunk 3]
```diff
@@ -178,7 +179,7 @@ export const useRfoxBridge = ({ confirmedQuote }: UseRfoxBridgeProps): UseRfoxBr
   }, [assetsById])
 
   const tradeQuoteInputWithWallet = useMemo(
-    () => ({ ...tradeQuoteInput, wallet }),
+    () => ({ ...tradeQuoteInput, wallet, quoteOrRate: 'quote' as const }),
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/pages/RFOX/components/Stake/Bridge/hooks/useRfoxBridge.ts:182]: Happy to cleanup as follow-up 🙏🏽 unrelated to the split per se

File: packages/swapper/src/types.ts [hunk 4]
```diff
@@ -425,6 +459,10 @@ export type QuoteResult = Result<TradeQuote[], SwapErrorRight> & {
   swapperName: SwapperName
 }
 
+export type RateResult = Result<TradeRate[], SwapErrorRight> & {
+  swapperName: SwapperName
+}
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/swapper/src/types.ts:464]: Happy to cleanup as follow-up 🙏🏽 unrelated to the split per se

**Review by 0xApotheosis — COMMENTED** [review]
Holy moly, monster 🐐 PR ser.

Code-wise looking excellent and tidy. Non-blocking, but it might be better if the `quoteOrRate` logic was an enum (though this may have implications on type narrowing).

Working on the functional review as we speak.

**Review by 0xApotheosis — APPROVED** [review]
✅ Able to get quotes for (and execute) permit2 ZRX with permit2 flag

Screenshot 2024-10-31 at 19 50 16

✅ Able to get quotes for (and execute) classic ZRX without permit2 flag

Screenshot 2024-10-31 at 20 01 15

✅ Able to get quotes for (and execute) Portals (note, this is a non-breaking change, hence the super hacky "quote then maybe estimate then maybe another dummy quote" flow is still present, to be removed when we wire this up with rates instead of quotes, with "TBA" for network fees)

Screenshot 2024-10-31 at 19 54 02

✅ Able to get quotes (and execute) for CoW

Screenshot 2024-10-31 at 19 56 24

✅ Able to get quotes (and execute) for Li.Fi

Screenshot 2024-10-31 at 19 52 05

✅ Able to get quotes (and execute) for THOR

Screenshot 2024-10-31 at 19 57 55

In addition, _rates_ are happy with `REACT_APP_FEATURE_PUBLIC_TRADE_ROUTE` - note, I did not attempt to execute any of these.

**Comment by 0xApotheosis** [comment]
I'll merge this beast in the name of progress. The outstanding nitpicks/non-blockers can be tackled in a future PR 🙏
